### PR TITLE
More net building utilities

### PR DESCRIPTION
**Pull request by emailweixu**
Also
* change alf.nn.Sequential to follow the interface of torch.nn.Sequentil
* Improved PredictiveRepresentationLearner for supporting post_processor and non-replay buffer based training when num_unrolll_steps==0

### CODE REVIEW

File: alf/networks/containers.py [hunk 1]
```diff
@@ -42,19 +41,25 @@ class Sequential(Network):
...
         new_state = (new_state1, new_state2)
 
-    """
-
-    def __init__(self, modules, input_tensor_spec=None, name="Sequential"):
-        """
-
         Args:
-            modules (list[nn.Module]): list of ``torch.nn.Module`` or ``alf.nn.Network``.
+            modules (nn.Module): list of ``torch.nn.Module`` or ``alf.nn.Network``.
```
Comment (Haichao-Zhang) [inline alf/networks/containers.py:45]: The comment in (), i.e. ``(nn.Module) `` should be updated?

File: alf/algorithms/predictive_representation_learner.py [hunk 2]
```diff
@@ -210,6 +216,8 @@ def __init__(self,
             observation_spec (nested TensorSpec): describing the observation.
             action_spec (nested BoundedTensorSpec): describing the action.
             num_unroll_steps (int): the number of future steps to predict.
+                if ``num_unroll_steps`` of 0 means no future prediction and hence
```
Comment (Haichao-Zhang) [inline alf/algorithms/predictive_representation_learner.py]: can remove ``if``

File: alf/layers.py [hunk 3]
```diff
@@ -2122,3 +2122,67 @@ def __init__(self):
...
+
+class Sequential(nn.Sequential):
+    """A thin wrapper over nn.Sequential to allow stateless alf.networks.Network."""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        for m in self:
+            if isinstance(m, alf.networks.Network):
+                assert not alf.nest.flatten(m.state_spec), (
+                    "Network element of layers.Sequential should stateless. "
```
Comment (Haichao-Zhang) [inline alf/layers.py]: should be

File: alf/algorithms/predictive_representation_learner.py [hunk 4]
```diff
@@ -230,6 +238,9 @@ def __init__(self,
                 dynamics net. Otherwise, a linear projection will be used to
                 convert the current latent represenation to the initial state for
                 the dynamics net.
+            postprocessor (None|Callable): If provided, will be called as
+                ``postprocessor(latent)`` to get the actual representation,
+                where ``latent`` is the output from encoding_net.
```
Comment (Haichao-Zhang) [inline alf/algorithms/predictive_representation_learner.py:246]: Is it useful to provide an option to specify postprocessor optimizer, to be structurally similar to ``encoding_net`` and ``dynamics net``? Might not be necessary in most of the cases when view postprocessor as part of the downstream networks, but might be useful in some cases.

File: alf/algorithms/predictive_representation_learner.py [hunk 5]
```diff
@@ -139,16 +139,21 @@ def calc_loss(self, target, predicted, mask=None):
             target = self._target_normalizer.normalize(target)
             predicted = self._target_normalizer.normalize(predicted)
 
-        loss = self._loss(predicted, target)
+        b = predicted.shape[0] * predicted.shape[1]
+        loss = self._loss(
+            predicted.reshape(b, *predicted.shape[2:]),
+            target.reshape(b, *target.shape[2:]))
+        loss = loss.reshape(*predicted.shape[:2], *loss.shape[1:])
```
Comment (hnyu) [inline alf/algorithms/predictive_representation_learner.py:148]: Why bothering reshaping here? I believe the loss class can handle arbitrary shapes.

File: alf/layers.py [hunk 6]
```diff
@@ -2122,3 +2122,67 @@ def __init__(self):
...
+        super().__init__()
+        has_network = any(
+            alf.nest.flatten(
+                alf.nest.map_structure(
+                    lambda m: isinstance(m, alf.networks.Network), modules)))
+        assert not has_network, (
+            "modules should not contain alf.networks.Network. "
+            "Try alf.networks.Branch instead.")
+
+        self._networks = modules
```
Comment (hnyu) [inline alf/layers.py:2159]: It's confusing that the variable name is "networks" but in fact it can't contain any network

File: alf/networks/containers.py [hunk 7]
```diff
@@ -42,19 +41,25 @@ class Sequential(Network):
...
             input_tensor_spec (TensorSpec): the tensor spec of the input. It must
                 be specified if it cannot be inferred from ``modules[0]``.
             name (str):
-        """
+    """
+    # The reason that we use a wrapper function for _Sequential is that Network
+    # does not allow *args for __init__() (see _NetworkMeta.__new__()). And we
+    # want to use *modules here to make the interface consistant with
```
Comment (hnyu) [inline alf/networks/containers.py]: consistent

File: alf/networks/containers.py [hunk 8]
```diff
@@ -14,17 +14,16 @@
...
 import alf
 from alf.nest import (flatten, flatten_up_to, map_structure,
                       map_structure_up_to, pack_sequence_as)
+from alf.utils.spec_utils import is_same_spec
 from .network import Network, get_input_tensor_spec, wrap_as_network
 
 
-@gin.configurable
-class Sequential(Network):
+def Sequential(*modules, input_tensor_spec=None, name="Sequential"):
```
Comment (hnyu) [inline alf/networks/containers.py:26]: Should still keep "configurable"?

File: alf/networks/containers.py [hunk 9]
```diff
@@ -312,3 +318,118 @@ def copy(self, name=None):
...
+        state = (block_state, echo_output)
+
+    """
+
+    def __init__(self, block, input_tensor_spec=None):
+        """
+        Args:
+            block (Network): the module for performing the actual computation
+            input_tensor_spec (nested TensorSpec): If provided, it must match
+                the block.input_tensor_spec[0] or input_tensor_spec['input']
```
Comment (hnyu) [inline alf/networks/containers.py]: \``block.input_tensor_spec['input']\``

**Review by hnyu — CHANGES_REQUESTED** [review]

File: alf/algorithms/predictive_representation_learner.py (hunk 5 above)
Comment (emailweixu) [reply alf/algorithms/predictive_representation_learner.py:148]: CrossEntropyLoss does not behave as we expected. The shape of predicted is [T, B, N] and the shape of target is [T, B], where N is the number of categories. For this kind of multidimensional tensor, CrossEntropyLoss expect shapes like [T, N, B] and [T, B]

File: alf/algorithms/predictive_representation_learner.py (hunk 4 above)
Comment (emailweixu) [reply alf/algorithms/predictive_representation_learner.py:246]: added.

File: alf/algorithms/predictive_representation_learner.py (hunk 2 above)
Comment (emailweixu) [reply alf/algorithms/predictive_representation_learner.py]: removed

File: alf/layers.py (hunk 6 above)
Comment (emailweixu) [reply alf/layers.py:2159]: self._modules is already used nn.Module

File: alf/layers.py (hunk 3 above)
Comment (emailweixu) [reply alf/layers.py]: fixed

File: alf/networks/containers.py (hunk 8 above)
Comment (emailweixu) [reply alf/networks/containers.py:26]: As far as I know, none of existing gin files uses this. New configs should be written in python. So there is no need for it to be configurable.

File: alf/networks/containers.py (hunk 1 above)
Comment (emailweixu) [reply alf/networks/containers.py:45]: ``nn.Module`` in the parenthesis is the type of each module.

**Review by Haichao-Zhang — APPROVED** [review]

**Review by hnyu — APPROVED** [review]
